Prevent check

ronny2205 opened the pull request:
Adam and I were working on this task and found a way around updating the database if not necessary.
The method moving_into_check? in piece.rb works. We called it from move_to! and some tests are failing. I think the main issue is adjusting castling to the new situation - if castling causing check, we need to rollback, as we did in the new method.

jonathanpike commented:
I absolutely love that you're using `ActiveRecord::Transactions`!  Super smart move, @ronny2205 and @addstar34.  I was having trouble understanding why you were raising an `ActiveRecord::Rollback` exception (with the `fail` keyword, which I understand is synonymous to `raise`).  After some research, I found that an `ActiveRecord::Rollback` exception will roll back the database transaction without passing on the exception.  [See the API docs here](http://api.rubyonrails.org/classes/ActiveRecord/Rollback.html).

However, does your method need to return true, in terms of how your calling it in `move_to!`?  It doesn't seem to do this, although I could be missing something.

ronny2205 commented:
You are absolutely right @jonathanpike! I just tested it, using the 'true' you suggested and it jumps there. I didn't think it would jump there, so we were not sure what this method should return. But it works, thanks!

Fail is synonymous to raise. Rubocop likes 'fail' :)

jonathanpike commented:
Glad that worked :+1:  Doing this in a transaction is such a good idea! 

For an interesting perspective about why `fail` is preferred over `raise`, check out [this article from Avdi Grimm](http://devblog.avdi.org/2014/05/21/jim-weirich-on-exceptions/).

ronny2205 commented:
Thanks, I will!

ronny2205 commented:
Some thoughts about more changes that have to be done in order to check all the cases of moving
_into_check:


1. Inside the methods castle! and capture_en_passant!, we need to check for moving_into_check, and rolling back if so. (same as we did in the new method moving_into_check?)

2.  In move_to!: 
   After check_for_castling,  Return if castled.
(I think we can return if there was no casling as well, because it's not gonna be a regular valid move for the king)

takehiromouri commented on app/models/piece.rb:
```diff
@@ -11,14 +11,27 @@ def move_to!(new_row, new_col)
     # Checking for Valid Move
     return unless valid_move?(new_row, new_col)
+    # Checking if the piece is moving into check
+    return unless !moving_into_check?(new_row, new_col)
```
a minor thing, but I think we should change this to 
`return if moving_into_check?(new_row, new_col)` because it makes it slightly more readable :smile: 

In the Ruby Style Guide I think it mentioned this as well (avoid using `unless !some_method` and use `if some_method` instead)

ronny2205 replied:
Yes, you're right! The reason we did it that way, is becuse moving_into_check? didn't return true before. After @jonathanpike's comment it does.

takehiromouri replied:
Ohh I see! great!

takehiromouri commented:
Great job @ronny2205 @addstar34 ! I love your use of Transaction as well! 

### Inside the methods castle! and capture_en_passant!, we need to check for moving_into_check, and rolling back if so. (same as we did in the new method moving_into_check?)
* I agree with this. I think we can just call `moving_into_check?` inside the two methods

ronny2205 commented:
Hi everybody,
1. I made all the changes, but I ran into a problem:
two old tests of en-passant fail (in pawn_test.rb). I commented them for now.
After debugging for a while I think I found out the reason, but it's weird:
Inside the method 'check_en_passant' the value of the row_dest changes for unknown reason (I added a comment about that in line #41 in pawn.rb).

If somebody wants to take a look or solve it together, let me know.

2. I updated castling and en-passant  - checking for check.  I didn't use the method 'moving_into_check?' because it does only one update and here we have 2 different updates.
I could refactor 'moving_into_check?' so it can deal with all the cases, but I'm saving it for later if time permits.

takehiromouri commented:
@ronny2205 Since I wrote the code for those parts, I'd be happy to go over it with you :)

ronny2205 commented:
@takehiromouri, let's do that. Are you free, by chance, today around 12:30pm?
